src/main/java/uk/gov/pay/adminusers/service/SecondFactorAuthenticator.java

```diff
@@ -1,31 +1,55 @@
 package uk.gov.pay.adminusers.service;
 
+import com.google.inject.Inject;
 import com.warrenstrange.googleauth.GoogleAuthenticator;
 import org.slf4j.Logger;
 import uk.gov.pay.adminusers.logger.PayLoggerFactory;
 
+import java.time.Clock;
+
 import static com.google.common.io.BaseEncoding.base32;
 import static org.apache.commons.lang3.StringUtils.isBlank;
 
 
 public class SecondFactorAuthenticator {
 
     private static final Logger logger = PayLoggerFactory.getLogger(SecondFactorAuthenticator.class);
+    private static final int VALID_PAST_TIME_WINDOWS = 4; // Plus 1 and then times 30 for maximum validity
+
 
     private final GoogleAuthenticator authenticator;
+    private final Clock clock;
 
-    public SecondFactorAuthenticator() {
+    @Inject
+    public SecondFactorAuthenticator(Clock clock) {
+        this.clock = clock;
         this.authenticator = new GoogleAuthenticator();
     }
 
     public int newPassCode(String secret) {
         checkNull(secret);
-        return authenticator.getTotpPassword(base32().encode(secret.getBytes()));
+        return authenticator.getTotpPassword(base32().encode(secret.getBytes()), clock.millis());
     }
 
+
+    // todo: make this more performant by checking out the minimum amount of windows needed to validate against all of windows in the range specified
     public boolean authorize(String secret, int passcode) {
         checkNull(secret);
-        return authenticator.authorize(base32().encode(secret.getBytes()), passcode);
+        long currentTime = clock.millis();
+        /*
+         * Validate against the current 30 second time window and the amount of past windows defined in 'VALID_PAST_TIME_WINDOWS'
+         * The authenticator already checks against windows either side of the current one,
+         * Thus we only want to check against 1 less than 'VALID_PAST_TIME_WINDOWS',
+         * As the last one will be checked automatically by the authenticator
+         */
+        for (int i = 0; i < VALID_PAST_TIME_WINDOWS; i++) {
+            long windowTime = currentTime - (30000L * i);
```

Comment:
I think hardcoding the 30000L here is a bit problematic? The 30 is actually what ever the value in config is from the GoogleAuthenticatorConfig https://github.com/wstrange/GoogleAuth/blob/master/src/main/java/com/warrenstrange/googleauth/GoogleAuthenticatorConfig.java#L36